File: 2-1.js
const shapes = {
    A: "rock",
    B: "paper",
    C: "scissors",
    X: "rock",
    Y: "paper",
    Z: "scissors",
};

const outcomes = {
    rock: {
        rock: "draw",
        paper: "loss",
        scissors: "win",
    },
    paper: {
        rock: "win",
        paper: "draw",
        scissors: "loss",
    },
    scissors: {
        rock: "loss",
        paper: "win",
        scissors: "draw",
    },
};

const points = {
    rock: 1,
    paper: 2,
    scissors: 3,
    loss: 0,
    draw: 3,
    win: 6,
};

require('fs').readFile("2.txt", "utf-8", (err, data) => {
    const score = data
        .split("\n")
        .filter(line => line)
        .map(line => {
            const [theirs, ours] = line
                .split(" ")
                .map(c => shapes[c]);

            const outcome = outcomes[ours][theirs];
            return points[outcome] + points[ours];
        })
        .reduce((total, round) => total + round, 0);

    console.log(score);
});

File: 2-1.c
#include "prelude.h"

static char const example[] = "A Y\n"
			      "B X\n"
			      "C Z\n";

static int const points[3][3] = {
	{
		3 + 1,
		6 + 2,
		0 + 3,
	},
	{
		0 + 1,
		3 + 2,
		6 + 3,
	},
	{
		6 + 1,
		0 + 2,
		3 + 3,
	},
};

static int solve(char *input) {
	char theirs, ours;
	int score = 0;

	LINES(line, input) {
		if (!strcmp(line, "")) break;
		sscanf(line, "%c %c", &theirs, &ours);
		score += points[(int)(theirs - 'A')][(int)(ours - 'X')];
	}

	return score;
}

int main() {
	RUN(solve, strdup(example));
	RUN(solve, read_to_string("2.txt"));
	return 0;
}
